### PR TITLE
Integration Test Refactor

### PR DESCRIPTION
**Pull request by matt-boris**
For https://github.com/mozilla-services/kinto-dist/issues/1927

### CODE REVIEW

File: .circleci/config.yml
Comment (matt-boris) [inline .circleci/config.yml:44]: What do you guys think?

File: tests/pytest.ini
Comment (grahamalama) [inline tests/pytest.ini:3]: Since we use `setup.cfg` as the kitchen sink of configuration, should we add this config there instead of this new `pytest.ini` file?

File: tests/conftest.py
Comment (grahamalama) [inline tests/conftest.py:103]: Do you think it would be good to split this into separate fixtures, or at least not have this fixture generate three clients by default? If I read the test cases correctly, I think there's only one test case where we need more than the User client. Also, creating all three users each time we need just one client seems like an unexpected side effect.

Maybe we could use the [Factory as fixture](https://docs.pytest.org/en/latest/how-to/fixtures.html#factories-as-fixtures) pattern here. So the fixture would be something like 
```python
@pytest.fixture
def make_client():
    def _make_client(auth):
        ...
    return _make_client
```
then in tests, we can do:
```python
def test_with_client(make_client, auth):
    client = make_client(auth)
    ...

def test_with_clients(make_client, auth, editor_auth, reviewer_auth):
    client = make_client(auth)
    editor_client = make_client(editor_auth)
    reviewer_client = make_client(reviewer_auth)
    ...
```

**Review by grahamalama — COMMENTED** [review]
Looking good so far! Just an initial round of comments. Mostly inline, but I was wondering if there was a reason in particular you chose to remove the `$` from shell code blocks in the README?

File: tests/conftest.py
Comment (matt-boris) [reply tests/conftest.py:103]: This is awesome, Graham! Thanks for the great idea

File: docker-compose.yml
Comment (matt-boris) [reply docker-compose.yml:9]: The [Docker](https://marketplace.visualstudio.com/items?itemName=ms-azuretools.vscode-docker) VS Code extension comes with a formatter that I use.

File: tests/integration_test.py
Comment (leplatrem) [reply tests/integration_test.py:331]: Good catch

File: .circleci/config.yml
Comment (leplatrem) [reply .circleci/config.yml:44]: I understand that the initial idea is to avoid pushing an image on Docker if the integration tests fail on it, but there must be a way to express that in the CircleCI config using two distinct jobs?

File: config/example.ini
Comment (leplatrem) [inline config/example.ini:135]: On our servers (DEV and STAGE), this would be `/buckets/main-workspace/collections/product-integrity`

File: tests/conftest.py
Comment (leplatrem) [inline tests/conftest.py:19]: I think we could make these tests closer to our current Remote Settings configuration. We could use `bucket=main-workspace` and `collection=product-integrity`.

The idea is to be able (eventually) to run this test suite against a freshly deployed instance.  But we can work towards that and adjust a few things in a follow-up PR

File: tests/integration_test.py
Comment (leplatrem) [inline tests/integration_test.py:203]: Maybe we can remove this (see #1946). We could assume that we are testing the usual Remote Settings setup `main-workspace/{cid} -> main-preview/{cid} -> main/{cid}`

File: Makefile
Comment (leplatrem) [inline Makefile:33]: nit: would `integration-test` be too long?

File: tests/integration_test.py
Comment (leplatrem) [inline tests/integration_test.py:359]: I believe could set the project_name settings to contain `Remote Settings` or this won't pass in DEV/STAGE

File: README.rst
Comment (leplatrem) [inline README.rst:74]: Nit: in kinto-http.py we have `make runkinto` https://github.com/Kinto/kinto-http.py/blob/bff433440a1eb468ff4d336349aada20fb73b357/Makefile#L32
but in kinto-attachment we have `make run-kinto` https://github.com/Kinto/kinto-attachment/blob/2d58f8a341073602089f97087c792b0eea836014/Makefile#L53

Take advantage of this to choose the one you prefer and we align them all :)

File: tests/integration_test.py
Comment (leplatrem) [inline tests/integration_test.py:19]: ditto about using `product-integrity` here too :)

File: tests/integration_test.py
Comment (leplatrem) [inline tests/integration_test.py:348]: Wouldn't we want to have this in a separate file? (eg. `browser_test.py` ?)

**Review by leplatrem — COMMENTED** [review]
Nice :clap: 

I like the shape it's taking!

Most of my comments are not mandatory to take into account. My main concern would be to adjust some of the settings / buckets / collection to values that reflect our Remote Settings instances. But you took the tests in the state they were and it makes sense. We could tackle this part in a follow-up if necessary.

Regarding naming and structure, we could spend some time to make sure it looks the way we 
want. We will now have:
- `kinto_remote_settings/signer/tests/test_*.py` : signer unit tests
- `kinto_remote_settings/changes/tests/test_*.py` : changes unit tests
- `kinto_remote_settings/signer/tests/functional.py` : signer/history tests against a real server
- `tests/integration_test.py` : tests of the integrated «service» + browser tests

I have the intuition that we could merge the last two (please open a follow-up issue if you agree). 

Kudos on this, great job! :muscle: 

> Please do not force push or rebase in order to ease the next review phases :pray:

File: pytest.ini
Comment (matt-boris) [reply pytest.ini]: https://github.com/mozilla-services/kinto-dist/pull/1953/commits/fc71b694dde9fc57aab13067763dc6dba3658932#diff-fa602a8a75dc9dcc92261bac5f533c2a85e34fcceaff63b3a3a81d9acde2fc52R20-R26

File: tests/pytest.ini
Comment (matt-boris) [reply tests/pytest.ini]: https://github.com/mozilla-services/kinto-dist/pull/1953/commits/fc71b694dde9fc57aab13067763dc6dba3658932#diff-fa602a8a75dc9dcc92261bac5f533c2a85e34fcceaff63b3a3a81d9acde2fc52R20-R26

File: tests/integration_test.py
Comment (matt-boris) [reply tests/integration_test.py]: https://github.com/mozilla-services/kinto-dist/pull/1953/commits/574f109903fd2b5ceba598f0a0e10b8eb219de5e

File: .flake8
Comment (matt-boris) [reply .flake8]: https://github.com/mozilla-services/kinto-dist/pull/1953/commits/fc71b694dde9fc57aab13067763dc6dba3658932

File: README.rst
Comment (matt-boris) [reply README.rst:74]: https://github.com/Kinto/kinto-http.py/pull/272

File: README.rst
Comment (matt-boris) [reply README.rst]: Circle CI has buildkit enabled already, so I just decided to link to a Docker support page on enabling that feature to help devs run the tests. https://github.com/mozilla-services/kinto-dist/pull/1953/commits/ea7f6641ab93887366e5b009d1a2c85843383af8

File: README.rst
Comment (matt-boris) [reply README.rst]: https://github.com/mozilla-services/kinto-dist/pull/1953/commits/07b37b4a858c422a346404a48bfef638e4eb0f78

File: tests/conftest.py
Comment (matt-boris) [reply tests/conftest.py]: https://github.com/mozilla-services/kinto-dist/pull/1953/commits/d9cb5bf0984244f69fb09bb42a9eb6f8846f7615

File: tests/integration_test.py
Comment (matt-boris) [reply tests/integration_test.py]: @leplatrem this has been updated now, will you please re-review the changes endpoint test?

File: tests/integration_test.py
Comment (matt-boris) [reply tests/integration_test.py:209]: Fixed with https://github.com/mozilla-services/kinto-dist/pull/1953/commits/d9cb5bf0984244f69fb09bb42a9eb6f8846f7615!

File: tests/integration_test.py
Comment (matt-boris) [reply tests/integration_test.py]: :) https://github.com/mozilla-services/kinto-dist/pull/1953/commits/a4a1553d5805b1cea57c5fc9dbb36d3b13cc345a

File: tests/integration_test.py
Comment (matt-boris) [reply tests/integration_test.py]: https://github.com/mozilla-services/kinto-dist/pull/1953/commits/ae50147473252fcc96d92271340469a55e9e8109

File: .circleci/config.yml
Comment (matt-boris) [reply .circleci/config.yml]: Got it with https://github.com/mozilla-services/kinto-dist/pull/1953/commits/a41b6e506d9ee6f1fbde4a6598e80d0293b9a639

Screen Shot 2021-12-14 at 3 55 30 PM

File: tests/integration_test.py
Comment (matt-boris) [reply tests/integration_test.py]: https://github.com/mozilla-services/kinto-dist/pull/1953/commits/9644e09acba2074bec8483b5b356b9bbc01da5bd

Will add more tests.

File: tests/integration_test.py
Comment (matt-boris) [reply tests/integration_test.py]: No longer checking the title content: https://github.com/mozilla-services/kinto-dist/pull/1953/commits/9644e09acba2074bec8483b5b356b9bbc01da5bd#diff-03ae4d94c3127061bcc8c5d44da8fe1745f8fe0ff6d3018e3cd95f6c697b3df6R17-R18

**Comment by matt-boris** [comment]
Leaving in draft for now, as I'd like to add some more browser tests for the `admin` plugin.

**Review by grahamalama — COMMENTED** [review]
Another round of minor comments 🔍

File: .gitignore
Comment (matt-boris) [reply .gitignore:73]: So I needed a way for the tests to get the _main_ `setup.cfg` file for the "kitchen sink" config, so I added this https://github.com/mozilla-services/kinto-dist/pull/1953/files#diff-e45e45baeda1c1e73482975a664062aa56f20c03dd9d64a827aba57775bed0d3R68. Because of that, it'll write the file to the `tests` directory as well and I didn't want two of the same file, so I added to `.gitignore`.

If there's a better way to go about doing this so the tests get that config file, I'm all ears!

File: Makefile
Comment (matt-boris) [reply Makefile:46]: Good call! I'll look into this.

File: Makefile
Comment (matt-boris) [reply Makefile:38]: I suppose I could try checking the `/v1/__heartbeat__` endpoint to ensure we're getting back the expected fields?

File: kinto_remote_settings/signer/tests/functional.py
Comment (matt-boris) [reply kinto_remote_settings/signer/tests/functional.py:137]: Yeah.... this one pained me to change to this. I probably should've just dropped the whole line in the first place, but I was just going through and replacing the `collection_timestamp()` calls like https://github.com/mozilla-services/kinto-dist/pull/1953/files/9644e09acba2074bec8483b5b356b9bbc01da5bd#diff-dcfbdbfdc3ed9ce49dc9a7df51c07335728904b744c2a7d6b966b080260e6c04L141 (which looks to be doing the same thing??). I can definitely remove this.

File: tests/conftest.py
Comment (matt-boris) [reply tests/conftest.py:6]: 🤷🏻‍♂️ it's definitely nice having the type hints/autocomplete in the editor. I'll read into some standards/suggestions and will post back here what I find. I'm not sure _how bad_ it is importing the private package.

File: tests/conftest.py
Comment (matt-boris) [reply tests/conftest.py:174]: This is awesome! I did not know that. Thanks

File: Makefile
Comment (matt-boris) [reply Makefile:46]: https://github.com/mozilla-services/kinto-dist/pull/1953/commits/a49142be70236c0f77277d7407095c12a355558f

File: Makefile
Comment (grahamalama) [reply Makefile:38]: I thought about this some more. Since we run everything through Docker and docker-compose, maybe we can bake this check into the command we set for the `tests` service instead of having this `need-kinto-running` step. The Docker docs have some suggestions of how to accomplish this [here](https://docs.docker.com/compose/startup-order/).

The nice thing about this approach is that we'd be able to target the service running at `web:8888`, so we'll know that it's our kinto service.

Comment (matt-boris) [reply Makefile]: Seems like the docker compose file was already pretty much set for this to work, just needed to add an additional step in the integration test make command. https://github.com/mozilla-services/kinto-dist/pull/1953/commits/46ba22e4160f1f37c229065fe594aef695b87789

File: kinto_remote_settings/signer/tests/functional.py
Comment (matt-boris) [reply kinto_remote_settings/signer/tests/functional.py]: https://github.com/mozilla-services/kinto-dist/pull/1953/commits/f6f890333b1c95914216d37837cbba440a673b2e

File: tests/conftest.py
Comment (matt-boris) [reply tests/conftest.py]: So I had just taken the recommended import path from vs code, which was incorrect of me to do since `FixtureRequest` is actually available from the public `pytest` package 🤦🏻‍♂️  https://github.com/pytest-dev/pytest/blob/ee93557ef33c03d7a792af1921806b54fe7c7a89/src/pytest/__init__.py#L99

Fixed in https://github.com/mozilla-services/kinto-dist/pull/1953/commits/df2c1d2b2afc3f162c63e4aa3cb0a8c223003f37

**Comment by grahamalama** [comment]
I'm happy with the Python aspect of this PR! I'll let @leplatrem look at how well these tests capture the integration checks we need, but otherwise LGTM 🚢.

File: tests/integration_test.py
Comment (leplatrem) [inline tests/integration_test.py:187]: nit: not sure about these prints anymore

File: tests/integration_test.py
Comment (leplatrem) [inline tests/integration_test.py:197]: cf. above, bucket perms redundant

**Review by leplatrem — APPROVED** [review]
This is great :100: :muscle: 

I left a few comments to be addressed ideally, but otherwise ready to merge  :tada: Great job !

**Review by leplatrem — APPROVED** [review]
:shipit:

**Pull request merged**
